### PR TITLE
[COTCO-559] Sanjay restricted to 1 queue

### PR DESCRIPTION
**Pull request by robertbutacu**

### CODE REVIEW

File: app/uk/gov/hmrc/ctc/errors/ErrorResponse.scala [hunk 1]
```diff
@@ -30,6 +30,8 @@ case class MongoDBError(message: String) extends CTCError
 case class MongoDBNoResults(message: String) extends CTCError
 case class DuplicatedMRNRecord(message: String) extends CTCError
 case class TimeoutExtensionNotApplicable(transitId: TransitId, message: String) extends CTCError
+case class UserAlreadyAssignedMrn(message: String = "User already assigned MRN") extends CTCError
+case class MissingProviderId(message: String = "Missing provider id") extends CTCError
```
[the remark this answers is not in the document]
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/errors/ErrorResponse.scala]: No

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 2]
```diff
@@ -46,26 +44,88 @@ class TransitService @Inject()(transitRepository: TransitRepository, appConfig:
...
+          result <- transitRepository.complete(id, providerId, record).map(_.right.map(result => result.id))
+          _      <- staffRepository.removeAllByProviderId(id)
+        } yield result
+    }
   }
 
   def release()(implicit hc: HeaderCarrier): ServiceResponse[Int] = {
     //We should alert or log if the result is greater than 1
-    transitRepository.releaseTransitsFor(providerId(hc))
+    withProviderId {
```
[the remark this answers is not in the document]
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: its going to be the same amount of code, but on the controller side

File: app/uk/gov/hmrc/ctc/repositories/StaffRepository.scala [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  def removeExpired(providerId: String): Future[Int] = {
+    val withProviderId: (String, JsValueWrapper) = "providerId" -> providerId
+    val withExpiredRecord: (String, JsValueWrapper) = "expirationTime" -> Json.obj("$lt"  -> Instant.now.toEpochMilli)
+
+    remove(withProviderId, withExpiredRecord).map(r => r.n)
+  }
+
+  def extendTimeout(transitId: String, providerId: String, timeoutExtension: Int): Future[Unit] = {
```
[the remark this answers is not in the document]
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/repositories/StaffRepository.scala]: Yes, it is important - relying on the client to always make sure he releases/completes processing an MRN is not a good idea. If, for some reason, a client decides to pick up the next MRN, but doesn't release/complete, then that its going to stay in the collection. Next time when he decides to pick up another MRN, that expired record is going to be removed.

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 4]
```diff
@@ -46,26 +44,88 @@ class TransitService @Inject()(transitRepository: TransitRepository, appConfig:
...
+          result  <- updateStaffRepository(nextMrn, providerId)
+        } yield result.right.map(transit => TransitData(transit, inactivity))
+      }
+    }
+
+    withProviderId {
+      providerId =>
+        for {
+          _        <- staffRepository.removeExpired(providerId)
+          mrnCount <- staffRepository.removeAllByProviderId(providerId)
```
[the remark this answers is not in the document]
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/services/TransitService.scala]: As states on Slack, I suggest reading the ticket.

File: app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala [hunk 5]
```diff
@@ -24,18 +24,20 @@ import uk.gov.hmrc.ctc.errors._
...
   def toResult(error: CTCError): Result = {
     error match {
       case e: TimeoutExtensionNotApplicable => Forbidden(Json.toJson(ErrorResponse(403, e.message)))
       case e: MalformedRequest => BadRequest(Json.toJson(ErrorResponse(400, e.message)))
       case _: MongoDBNoResults => NotFound(Json.toJson(ErrorResponse(404, "No results found for the request")))
-      case _                   => InternalServerError(Json.toJson(ErrorResponse(500, "The request can't be processed")))
+      case _: MongoDBError     => NotFound(Json.toJson(ErrorResponse(404, "No results found for the request")))
+      case e: UserAlreadyAssignedMrn => Conflict(Json.toJson(ErrorResponse(409, e.message)))
+      case e: MissingProviderId => BadRequest(Json.toJson(ErrorResponse(400, e.message)))
+      case e => InternalServerError(Json.toJson(ErrorResponse(500, e.message)))
```
[the remark this answers is not in the document]
Comment (robertbutacu) [reply app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala]: This is an exception defined by us - so its rather up to us to not have the stack trace as the message. But having the static value is better, indeed.

File: conf/application.conf [hunk 6]
```diff
@@ -32,7 +32,7 @@ play.http.filters = "uk.gov.hmrc.play.bootstrap.filters.MicroserviceFilters"
 
 play.http.errorHandler = "uk.gov.hmrc.play.bootstrap.http.JsonErrorHandler"
 
-play.http.router = prod.Routes
+play.http.router = testOnlyDoNotUseInAppConf.Routes
```
Comment (peppe-hmrc) [inline conf/application.conf]: OPS :) ... btw you can use the command in the Readme to run test routes locally ;)

Comment (robertbutacu) [reply conf/application.conf]: always get this - too lazy to run the extra parameters... should maybe make an alias hahh

File: app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala [hunk 7]
```diff
@@ -25,18 +25,19 @@ import uk.gov.hmrc.ctc.errors._
...
   def toResult(error: CTCError): Result = {
     error match {
-      case e: TimeoutExtensionNotApplicable => Forbidden(Json.toJson(ErrorResponse(403, e.message)))
-      case e: MalformedRequest => BadRequest(Json.toJson(ErrorResponse(400, e.message)))
+      case e: TimeoutExtensionNotApplicable => Forbidden(Json.toJson(ErrorResponse(403, "Cannot extend timeout")))
+      case e: MalformedRequest => BadRequest(Json.toJson(ErrorResponse(400, "The request is incomplete")))
       case _: MongoDBNoResults => NotFound(Json.toJson(ErrorResponse(404, "No results found for the request")))
-      case _                   => InternalServerError(Json.toJson(ErrorResponse(500, "The request can't be processed")))
+      case _: MongoDBError     => NotFound(Json.toJson(ErrorResponse(404, "No results found for the request")))
+      case _: UserAlreadyAssignedTransit => Forbidden(Json.toJson(ErrorResponse(403, "User already assigned mrn")))
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/handlers/ErrorHandler.scala]: `"User is already processing a transit"`

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 8]
```diff
@@ -16,62 +16,99 @@
...
+import uk.gov.hmrc.ctc.repositories.{OperatorRepository, TransitRepository}
 import scala.concurrent.{ExecutionContext, Future}
 
 @Singleton
-class TransitService @Inject()(transitRepository: TransitRepository, appConfig: AppConfig)
+class TransitService @Inject()(transitRepository: TransitRepository,
+                               operatorRepository: OperatorRepository,
+                               appConfig: AppConfig)
                               (implicit ec: ExecutionContext) {
+  def submitTransit(transitRecord: Transit): ServiceResponse[TransitDocument] = {
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/services/TransitService.scala]: `def submit(transit: Transit)`

File: app/uk/gov/hmrc/ctc/controllers/actions/IsUserBlocked.scala [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+import play.api.mvc.Results._
+import uk.gov.hmrc.ctc.errors.ErrorResponse
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class IsUserBlocked @Inject()(transitService: TransitService)(implicit ec: ExecutionContext) extends ActionFilter[RequestWithProviderId]{
+  override protected def filter[A](request: RequestWithProviderId[A]): Future[Option[Result]] = {
+    transitService.isOperatorBlocked(request.providerId)
+      .map {
+        case Left(_)  => Some(Forbidden(Json.toJson(ErrorResponse(403, "Operator already assigned transit record"))))
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/controllers/actions/IsUserBlocked.scala]: `"Operator is already processing a transit"`

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 10]
```diff
@@ -16,62 +16,99 @@
...
+  def isOperatorBlocked(providerId: String): ServiceResponse[Unit] = operatorRepository.isOperatorBlocked(providerId)
+  //TODO StaffRepository could be the source of truth for InProgress transits, not having to go to transitRepository also
+  def nextTransit(destinationPort: DestinationPort, transitState: TransitSlot, providerId: String): ServiceResponse[TransitTimeout] = {
     val inactivity = appConfig.transitInactivityTimeout
-    val pid = providerId(hc)
-    transitRepository.getNextMrn(destinationPort, inactivity, pid ,transitState).map(
-      _.right.map(transitDocument => TransitTimeout(transitDocument, inactivity))
-    )
+
+    def updateOperatorRepository(record: Either[CTCError, TransitDocument], providerId: String): Future[Either[CTCError, TransitDocument]] = {
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/services/TransitService.scala]:
```
def updateOperatorRepository(
  response: Response[TransitDocument], providerId: String
): ServiceResponse[TransitDocument]
```

File: app/uk/gov/hmrc/ctc/services/TransitService.scala [hunk 11]
```diff
@@ -16,62 +16,99 @@
...
+      record match {
+        case l@Left(_) => Future.successful(l)
+        case Right(r)  =>
+          val operatorRecord = Operator.fromTransit(r, providerId, inactivity)
+
+          operatorRepository.insert(operatorRecord).map(_ => Right(r)) // backtrack if this fails?
+      }
+    }
+
+    def getNextTransitRecord(isOperatorAssigned: Either[CTCError, Unit], providerId: String): ServiceResponse[TransitTimeout] = {
```
Comment (udinobi) [inline app/uk/gov/hmrc/ctc/services/TransitService.scala]: `def getNextTransit(isOperatorAssigned: Response[Unit]...`

**Comment by udinobi** [comment]
I think it might be better rebase before to move forward with the review.

**Comment by robertbutacu** [comment]
Rebase with what? Master has been merged

**Pull request merged**